Firestore/core/src/firebase/firestore/remote/serializer.cc

```diff
@@ -415,6 +454,43 @@ std::string Reader::ReadString() {
   return result;
 }
 
+void EncodeTimestamp(Writer* writer, const Timestamp& timestamp_value) {
+  google_protobuf_Timestamp timestamp_proto =
+      google_protobuf_Timestamp_init_zero;
+  timestamp_proto.seconds = timestamp_value.seconds();
+  timestamp_proto.nanos = timestamp_value.nanoseconds();
+  writer->WriteNanopbMessage(google_protobuf_Timestamp_fields,
+                             &timestamp_proto);
+}
+
+Timestamp DecodeTimestamp(Reader* reader) {
+  google_protobuf_Timestamp timestamp_proto =
+      google_protobuf_Timestamp_init_zero;
+  reader->ReadNanopbMessage(google_protobuf_Timestamp_fields, &timestamp_proto);
+
+  // The Timestamp ctor will assert if we provide values outside the valid
+  // range. However, since we're decoding, a single corrupt byte could cause
+  // this to occur, so we'll verify the ranges before passing them in since we'd
+  // rather not abort in these situations.
+  if (timestamp_proto.seconds < TimestampInternal::Min().seconds()) {
+    reader->set_status(Status(FirestoreErrorCode::DataLoss,
+                              "Input proto bytes cannot be parsed (timestamp "
```

Comment:
Nit: This error message is kind of tortured and it's a little weird that the most specific thing we know is parenthetical.

How about: "Invalid message: timestamp beyond the earliest supported date"

Reply:
Done (also two immediately below.)